server/schedule/filter/filters.go

```diff
@@ -370,64 +372,119 @@ func (f StoreStateFilter) Type() string {
-		return false
-	}
-	if f.TransferLeader &&
-		(store.IsDisconnected() ||
-			store.IsBlocked() ||
-			store.IsBusy() ||
-			opts.CheckLabelProperty(opt.RejectLeader, store.GetLabels())) {
-		return false
-	}
+func (f StoreStateFilter) isDown(opt opt.Options, store *core.StoreInfo) bool {
+	return store.DownTime() > opt.GetMaxStoreDownTime()
+}
 
-	if f.MoveRegion {
-		// only target consider the pending peers because pending more means the disk is slower.
-		if opts.GetMaxPendingPeerCount() > 0 && store.GetPendingPeerCount() > int(opts.GetMaxPendingPeerCount()) {
-			return false
-		}
+func (f StoreStateFilter) isOffline(opt opt.Options, store *core.StoreInfo) bool {
+	return store.IsOffline()
+}
 
-		if !f.filterMoveRegion(opts, false, store) {
-			return false
+func (f StoreStateFilter) isBlockLeaderTransfer(opt opt.Options, store *core.StoreInfo) bool {
+	return store.IsBlocked()
+}
+
+func (f StoreStateFilter) isDisconnected(opt opt.Options, store *core.StoreInfo) bool {
+	return !f.AllowTemporaryStates && store.IsDisconnected()
+}
+
+func (f StoreStateFilter) isBusy(opt opt.Options, store *core.StoreInfo) bool {
+	return !f.AllowTemporaryStates && store.IsBusy()
+}
+
+func (f StoreStateFilter) exceedRemoveLimit(opt opt.Options, store *core.StoreInfo) bool {
+	return !f.AllowTemporaryStates && !store.IsAvailable(storelimit.RemovePeer)
+}
+
+func (f StoreStateFilter) exceedAddLimit(opt opt.Options, store *core.StoreInfo) bool {
+	return !f.AllowTemporaryStates && !store.IsAvailable(storelimit.AddPeer)
+}
+
+func (f StoreStateFilter) tooManySnapshots(opt opt.Options, store *core.StoreInfo) bool {
+	return !f.AllowTemporaryStates && (uint64(store.GetSendingSnapCount()) > opt.GetMaxSnapshotCount() ||
+		uint64(store.GetReceivingSnapCount()) > opt.GetMaxSnapshotCount() ||
+		uint64(store.GetApplyingSnapCount()) > opt.GetMaxSnapshotCount())
+}
+
+func (f StoreStateFilter) tooManyPendingPeers(opt opt.Options, store *core.StoreInfo) bool {
+	return !f.AllowTemporaryStates &&
+		opt.GetMaxPendingPeerCount() > 0 &&
+		store.GetPendingPeerCount() > int(opt.GetMaxPendingPeerCount())
+}
+
+func (f StoreStateFilter) hasRejectLeaderProperty(opts opt.Options, store *core.StoreInfo) bool {
+	return opts.CheckLabelProperty(opt.RejectLeader, store.GetLabels())
+}
+
+// The condition table.
+// Y: the condition is temporary (expected to become false soon).
+// N: the condition is expected to be true for a long time.
+// X means when the condition is true, the store CANNOT be selected.
+//
+// Condition    Down Offline Tomb Block Disconn Busy RmLimit AddLimit Snap Pending Reject
+// IsTemporary  N    N       N    N     Y       Y    Y       Y        Y    Y       N
+//
+// LeaderSource X            X    X     X
+// RegionSource                                 X    X                X
+// LeaderTarget X    X       X    X     X       X                                  X
+// RegionTarget X    X       X          X       X            X        X    X
```

Comment:
if store is blocked, it still can be selected as region target?

Reply:
yes, block may be a bad name. It only means block leader transfer. I may have some other PR to change the name.

Reply:
To block state is marked only by `evict-leader-scheduler` and `grant-leader-scheduler` AFAIK.